### PR TITLE
Add location scanner

### PR DESCRIPTION
**Pull request by umairidris**
Currently only supports buckets.

Part of #2053.

### CODE REVIEW

File: google/cloud/forseti/scanner/scanners/location_scanner.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        """Retrieves the data for scanner.
+
+        Returns:
+            List[Resource]: resources to check for violations.
+        Raises:
+            ValueError: if resources have an unexpected type.
+        """
+        resources = []
+
+        resource_type_to_fn = {'bucket': bucket.Bucket.from_json}
```
Comment (dufton) [inline google/cloud/forseti/scanner/scanners/location_scanner.py]: Document what kind of function is expected here. Perhaps we could move this this out to somewhere the rule engine and scanner can both see, so there is a single source of truth for supported resources

File: tests/scanner/audit/location_rules_engine_test.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        )
+        rules_engine = get_rules_engine_with_rule(rule)
+        self.assertEqual(1, len(rules_engine.rule_book.resource_to_rules))
+
+    def test_find_violations_bucket_whitelist_no_violations(self):
+        rule = rule_tmpl.format(
+            mode='whitelist',
+            type='bucket',
+            locations=['eu*'],
+        )
```
Comment (dufton) [inline tests/scanner/audit/location_rules_engine_test.py]: One of the tests should cover exact matches (no wildcard), and multiple locations in a single rule

File: google/cloud/forseti/common/gcp_type/bucket.py [hunk 3]
```diff
@@ -47,6 +50,7 @@ def __init__(
             name (str): The bucket's unique GCP name, with the
                 format "buckets/{id}".
             display_name (str): The bucket's display name.
+            locations (List[str]): Locations this bucket resides in.
```
[the remark this answers is not in the document]
Comment (umairidris) [reply google/cloud/forseti/common/gcp_type/bucket.py]: Done.

File: google/cloud/forseti/common/gcp_type/resource.py [hunk 4]
```diff
@@ -229,6 +232,15 @@ def parent(self):
         """
         return self._parent
 
+    @property
+    def locations(self):
+        """Locations the resource resides in.
```
[the remark this answers is not in the document]
Comment (umairidris) [reply google/cloud/forseti/common/gcp_type/resource.py]: Done.

File: google/cloud/forseti/scanner/audit/location_rules_engine.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+
+class Mode(enum.Enum):
+    """Rule modes."""
+    WHITELIST = 'whitelist'
+    BLACKLIST = 'blacklist'
+
+
+class LocationRulesEngine(base_rules_engine.BaseRulesEngine):
+    """Rules engine for Liens."""
```
[the remark this answers is not in the document]
Comment (umairidris) [reply google/cloud/forseti/scanner/audit/location_rules_engine.py]: Done.

File: google/cloud/forseti/scanner/audit/location_rules_engine.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        for raw_resource in resources:
+            resource_ids = raw_resource.get('resource_ids')
+
+            if not resource_ids:
+                raise errors.InvalidRulesSchemaError(
+                    'Missing resource ids in rule {}'.format(rule_index))
+
+            resource_type = raw_resource.get('type')
+
+            if resource_type not in {'project', 'folder', 'organization'}:
```
[the remark this answers is not in the document]
Comment (umairidris) [reply google/cloud/forseti/scanner/audit/location_rules_engine.py]: Done.

File: google/cloud/forseti/scanner/audit/location_rules_engine.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+                    location_patterns=rule_def.get('locations'))
+
+    def find_violations(self, resource):
+        """Find lien violations in the rule book.
+
+        Args:
+            resource (Resource): The GCP resource to check locations for.
+                This is where we start looking for rule violations and
+                we move up the resource hierarchy (if permitted by the
+                resource's "inherit_from_parents" property).
```
[the remark this answers is not in the document]
Comment (umairidris) [reply google/cloud/forseti/scanner/audit/location_rules_engine.py]: Done.

File: google/cloud/forseti/scanner/audit/location_rules_engine.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+                to.
+            location_patterns (List[str]): Forseti-style patterns for locations.
+        """
+        self.name = name
+        self.index = index
+        self.mode = mode
+        self.applies_to = applies_to
+
+        loc_re_str = '|'.join([
+            regular_exp.escape_and_globify(loc_wildcard)
```
[the remark this answers is not in the document]
Comment (umairidris) [reply google/cloud/forseti/scanner/audit/location_rules_engine.py]: Done.

File: tests/scanner/audit/location_rules_engine_test.py (hunk 2 above)
Comment (umairidris) [reply tests/scanner/audit/location_rules_engine_test.py]: Done.

File: google/cloud/forseti/scanner/scanners/location_scanner.py (hunk 1 above)
Comment (umairidris) [reply google/cloud/forseti/scanner/scanners/location_scanner.py]: Done.

**Pull request merged**